Silence warnings on project settings and third party code in Xcode & resolve some warnings

montegoulding opened the pull request:
Unfortunately this required a minor patch to gyp which I'm not sure we want to modify if we haven't yet. The project update warnings are impossible to bypass any other way unfortunately.

The other thing that could be done is silence warnings on perfect unless we are patching that.

Commit:
Silence warnings on project settings and third party code in Xcode

Commit:
Add missing read_only initialiser

Commit:
Remove unused variables

Commit:
Remove unnecessary code and unused variable from charIndex

montegoulding commented on engine/src/exec-interface-field-chunk.cpp:
```diff
@@ -1538,15 +1538,7 @@ void MCField::GetEffectiveFormattedStyledTextOfCharChunk(MCExecContext& ctxt, ui
 // AL-2014-05-27: [[ Bug 12511 ]] charIndex is a char chunk property
```
@livecodeali here's some context to my posts on slack ;-)

peter-b commented:
@montegoulding Rather than modifying all of those initialisations, I think you can make the error go away just by giving `MCExecEnumTypeElementInfo` a `constexpr` constructor with default arguments, e.g.

``` c++
struct MCExecEnumTypeElementInfo
{
    constexpr MExecEnumTypeElementInfo(const char[] p_tag, intenum_t p_value, bool p_read_only=false)
        : tag(p_tag), value(p_value), read_only(p_read_only) {}
};
```

This would prevent the warnings without generating any additional code, since the initialisation would be performed at compile time.

montegoulding commented:
@peter-b that seems much more logical than what I've done thanks!

peter-b commented:
The only reason I can suggest it is that we now have C++11 compiler support across all platforms (I think)!

Commit:
Remove unused variables

Commit:
Fix kMCPickPhotoSourceTypeTypeInfo to use correct enum values

Commit:
Remove kMCParagraphAttrsProps

Commit:
Remove unused variables

Commit:
Implement default cases for unhanded enumeration values

A number of potentially uninitialised variables
were masked as enumeration value warnings rather
than uninitialised variable errors

Commit:
Fix dangling else warnings

montegoulding commented:
Removing from milestone for the time being as there's something in the last commit breaking things badly and it's not overly important which version it goes in.

Commit:
Fix variable declaration shadows variable warnings

montegoulding commented:
Ah... turned out that something that looked like an unnecessary re-declaration was in fact just a regular shadowed variable name and needed to be renamed

Commit:
Fix && within || warnings

Commit:
Fix variable declaration shadows variable warnings

Commit:
Fix delete on pointers allocated with new[]

Commit:
Fix widget custom callback initialiser warning

Commit:
Fix use of logical && with constant operand

Commit:
Remove extraneous parenthesis

Commit:
Remove unnecessary semicolons

Commit:
Implement default cases for unhanded enumeration values

montegoulding commented:
OK, that's it for the time being. Let me know if you want the PR divided up into separate patches so it's easier to review.

Commit:
Fix some comparison of unsigned to signed warnings

Commit:
Fix regression to vGrid line drawing

Commit:
Fix redefinition of min & max types

Commit:
Use MAXUINT4 instead of negative -1 assignment to unsigned

Commit:
Use CODEPOINT_NONE instead of assigning negatives to unsigned

Commit:
Fix direct comparison to NSNumber

Commit:
Silence warnings on integer clamp

Commit:
Fix bug in MCUnicodeGetProperty kMCUnicodePropertyTypeUint32

Commit:
Resolve comparison of different signed vars

peter-b commented on engine/src/mac-surface.mm:
```diff
@@ -396,6 +396,8 @@ static inline CGBlendMode MCGBlendModeToCGBlendMode(MCGBlendMode p_blend)
 		case kMCGBlendModeLuminosity:
 			return kCGBlendModeLuminosity;
+        default:
+            break;
```
I suggest you move the `MCUnreachableReturn()` into the `default:` case.

peter-b commented on engine/src/funcs.cpp:
```diff
@@ -578,6 +578,8 @@ void MCChunkOffset::compile(MCSyntaxFactoryRef ctxt)
 		MCSyntaxFactoryEvalMethod(ctxt, kMCStringsEvalOffsetMethodInfo);
 		break;
+    default:
+        break;
```
I'm pretty certain this should be `MCUnreachable()`. If reached it would indicate that the parser had accepted something that couldn't actually be compiled.

Commit:
Fix different signed variable comparisons

peter-b commented on engine/src/ibmp.cpp:
```diff
@@ -2902,6 +2908,8 @@ bool MCImageEncodeRawTrueColor(MCImageBitmap *p_bitmap, IO_handle p_stream, Expo
 			MCBitmapConvertRow<EX_RAW_ABGR>(t_row_buffer, (uint32_t*)t_src_ptr, p_bitmap->width);
 			break;
+        default:
+            break;
```
This should be `MCUnreachableReturn(false)`.  If you just break here then the following `IO_write()` will output an uninitialised buffer.

peter-b commented on engine/src/ide.cpp:
```diff
@@ -2545,6 +2545,8 @@ struct MCIdeFilterControlsVisitor: public MCObjectVisitor
             }
             break;
+            default:
+                break;
```
I'm pretty sure this is a case where the switch should explicitly handle _all_ possible members of the enumeration.

Commit:
Move unreachable return to default case

peter-b commented on engine/src/idraw.cpp:
```diff
@@ -317,6 +319,8 @@ void MCImage::drawwithgravity(MCDC *dc, MCRectangle r, MCGravity p_gravity)
             dx = r . x;
             dw = r . width;
             break;
+        default:
```
I think this should just be `case kMCGravityNone:` (and the same for the following `switch` block).

peter-b commented on engine/src/image.cpp:
```diff
@@ -413,6 +413,8 @@ Boolean MCImage::mdown(uint2 which)
 
 							return True;
 						}
+                    default:
```
I think you need a `break;` for the preceding set of `case`s to prevent them falling through into the `default` case.

peter-b commented on engine/src/osxtheme.mm:
```diff
@@ -182,6 +184,8 @@ static void converttomcrect(const Rect &macR, MCRectangle &mcR)
         return 21;
         break;
+    default:
+        break;
```
Move the `return 0;` into the default case.

peter-b commented on engine/src/osxtheme.mm:
```diff
@@ -152,6 +152,8 @@ static void converttomcrect(const Rect &macR, MCRectangle &mcR)
 		return True;
 		break;
+    default:
+        break;
```
Move the `return True;` into the default case.

peter-b replied:
Hang on this function is weird, it just returns `True` for everything.  Why not just change its whole body to `return True;`?

Otherwise surely its `switch` should cover all possible values of `Widget_Type` and return `False` for the ones not listed...?

montegoulding replied:
@peter-b the only reason I can think is documentation when comparing v other platorms theming but then you are right it should list all enum values... perhaps it was meant to return false?

peter-b commented on engine/src/osxtheme.mm:
```diff
@@ -409,7 +417,7 @@ static void converttomcrect(const Rect &macR, MCRectangle &mcR)
 {
 	MCRectangle trect = drect;
-	ThemeButtonKind themebuttonkind;
+	ThemeButtonKind themebuttonkind = 0;
```
This should be `kThemePushButton` please :)

peter-b commented on engine/src/srvcgi.cpp:
```diff
@@ -1816,6 +1816,8 @@ static bool cgi_send_headers(void)
 				sprintf(t_content_header, "Content-Type: text/html; charset=utf-8\n");
 				break;
+            default:
+                break;
```
I'm pretty sure this should be:

``` c++
case kMCSOutputTextEncodingUndefined:
    sprintf(t_content_header, "Content-Type: text/html\n");
    break;
```

peter-b commented on libgraphics/src/context.cpp:
```diff
@@ -2153,6 +2153,8 @@ static bool MCGContextApplyPaintSettingsToSkPaint(MCGContextRef self, MCGColor p
                 r_paint . setFilterLevel(SkPaint::kHigh_FilterLevel);
 				break;
+            default:
+                break;
```
In this case, the enumeration definition is wrong and this `default` case shouldn't be needed. `kMCGImageFilterCount` shouldn't be a member of the `MCGImageFilter` enumeration for exactly this reason.

montegoulding replied:
@peter-b hmm... ok, moving it to `module-canvas.cpp` which is the only place it's used and defining it as:

```
static const uint8_t kMCGImageFilterCount = 4;
```

peter-b replied:
@montegoulding Hmm, I'm worried that doing that puts the constant that describes the number of values in the enumeration a really long way from the definition of the enumeration.  I think it would probably be better to put the (non-`static`) definition of `kMCGImageFilterCount` as close as possible after the definition of the enumeration.  That will hopefully help make sure that if the `MCGImageFilter` enumeration is changed, the count gets changed too...

peter-b commented on libgraphics/src/legacygradients.cpp:
```diff
@@ -951,6 +951,9 @@ MCGradientAffineCombiner *MCGradientFillCreateCombiner(MCGGradientRef p_gradient
 		}
+            
+        default:
+            break;
```
See my previous comment about `MCGImageFilter`.  This default case shouldn't be needed.

peter-b commented on libfoundation/src/system-file-posix.cpp:
```diff
@@ -364,8 +364,6 @@ __MCSFileSetContents (MCStringRef p_native_path,
 	if (0 != rename (*t_temp_path_sys, *t_path_sys))
 	{
-		int t_save_errno = errno;
-
```
This change is incorrect; if you remove this then the message that's sent won't be the error code from `rename()`.  I suggest replacing with:

```
t_save_errno = errno;
```

montegoulding replied:
argh... sorry this is what I intended to do I think...

peter-b commented on engine/src/button.cpp:
```diff
@@ -1437,7 +1437,7 @@ Boolean MCButton::mup(uint2 which, bool p_release)
 	if (state & CS_GRAB)
 	{
-		if (flags && F_AUTO_HILITE)
+		if (flags & F_AUTO_HILITE)
```
Can you please turn this warning into an error in Mac builds?

montegoulding replied:
Yes, I'm planning on reviewing which errors I can turn on at the end and whether they can be turned on for all builds or just Mac.

peter-b commented on libgraphics/src/context.cpp:
```diff
@@ -2551,7 +2551,9 @@ static bool MCGContextDrawSkBitmap(MCGContextRef self, const SkBitmap &p_bitmap,
 					t_paint . setFilterLevel(SkPaint::kHigh_FilterLevel);
 					break;
-			}
+                default:
```
Please remove `kMCGImageFilterCount` instead.

peter-b commented on engine/src/imagebitmap.cpp:
```diff
@@ -361,11 +361,11 @@ void MCImageBitmapPremultiplyRegion(MCImageBitmap *p_bitmap, int32_t p_sx, int32
 	uint8_t *t_src_ptr = (uint8_t*)p_bitmap->data + t_src_x * 4 + t_src_y * p_bitmap -> stride;
 	uint8_t *t_dst_ptr = (uint8_t*)p_pixel_ptr + t_dst_x * 4 + t_dst_y * p_pixel_stride;
-	for (uindex_t y = 0; y < t_height; y++)
+	for (uindex_t y = 0; y < (uindex_t)t_height; y++)
```
In this case I recommend making `y` into a `index_t` rather than casting `t_height` to `uindex_t`.  `y` is never used as anything other than a counter in the `for` loop, and therefore it makes no difference whether `y` is signed or unsigned.

The same consideration applies to the `x` in the inner loop.

peter-b commented on engine/src/iutil.cpp:
```diff
@@ -977,14 +977,14 @@ bool MCImageRotateRotate(MCImageBitmap *p_src, real64_t p_angle, uint32_t p_back
 			else
 			{
-				if (t_ix < -1 || t_ix >= p_src->width || t_iy < -1 || t_iy >= p_src->height)
+				if (t_ix < -1 || (uint32_t)t_ix >= p_src->width || t_iy < -1 || (uint32_t)t_iy >= p_src->height)
```
Unfortunately this code appears to have been _originally_ written incorrectly.
1. If `t_ix == -1`, then `uint32_t(t_ix) == UINT32_MAX`.
2. Therefore if `t_ix == -1`, then `uint32_t(t_ix) >= p_src->width` for all values of `p_src->width`
3. Therefore if `t_ix == -1` then the true case of this `if` statement is _always_ executed.

However, there are `if t_ix == -1` checks in the false case of this `if` statement, which is tautological.

It's very unclear how to make this code correct.

montegoulding replied:
Hmm... do you think it should be `< 0` like the `INTERPOLATION_BOX` condition above?

peter-b replied:
I really have no idea.  I discussed with @livecodeali briefly before leaving the office yesterday afternoon and we're completely stumped.

livecodeali replied:
Doesn't help that it's been like that since the file was initially committed to git!

peter-b commented on engine/src/edittool.cpp:
```diff
@@ -38,14 +38,15 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 #include "graphic.h"
 #include "edittool.h"
 
+#define kInvalidEditPoint MAXUINT4
```
Because this is C++, you can safely define constants that have no runtime storage using e.g.

``` c++
static const uint4 kInvalidEditPoint = MAXUINT4;
```

I recommend this instead of `#define` for situations like this.  One interesting reason is that you get better debugging and error messages because the name is resolved by the compiler rather than the preprocessor.

peter-b commented on engine/src/mac-core.mm:
```diff
@@ -2056,7 +2056,7 @@ int platform_main(int argc, char *argv[], char *envp[])
 	/* UNCHECKED */ MCMemoryNewArray(1, t_new_envp);
 	
-	int i = 0;
+	uindex_t i = 0;
```
I'm confused.  `i` is used in this function in the loop:

``` c++
for (i = 0; i < argc; i++)
    MCValueRelease(t_new_argv[i]);
```

Now, `argc` is signed, so surely that introduces a signed-unsigned comparison?  Am I missing something?

I suggest introducing new separate loop variables, e.g. `t_env_num` and `t_arg_num`.

montegoulding replied:
@peter-b yes I think there was a distinct lack of brain applied by me on that

Commit:
Various fixes based on feedback

Commit:
Enable a number of errors

This commit enables:

- constant-logical-operand
- shadow
- parentheses
- objc-literal-compare

Additionally the following are close to being
able to be enabled:

- switch - issue with sqlite header
- sign-compare - issue with MCMin template

Commit:
Silence compact unwind warning

Commit:
MCCapusleInfo data initialiser warnings

Commit:
Unused variable warning

Commit:
Fix some conversion warnings

peter-b commented:
@montegoulding I'm afraid that I don't think the changes to the `BIO_write()` and `BIO_read()` invocations are quite right.  Those functions take an `int` as their length argument and return an `int` as their result.  To me, adding the explicit cast from the result to `uint32_t` seems like the wrong thing because it's masking the fact that we're using the API incorrectly in the first place.  I would be more comfortable with a patch that changes the types of the variables being used for computing and comparing the length parameters to `int`.

peter-b commented on engine/src/dskmac.cpp:
```diff
@@ -5540,9 +5540,9 @@ static bool startprocess_write_uint32_to_fd(int fd, uint32_t value)
+	if (!startprocess_write_uint32_to_fd(fd, uint32_t(strlen(string) + 1)))
 		return false;
-	if (write(fd, string, strlen(string) + 1) != strlen(string) + 1)
+    if (write(fd, string, strlen(string) + 1) != (ssize_t)strlen(string) + 1)
```
There are two problems here that adding these type casts have exposed
- What happens when trying to write a Very Big String on a 64-bit system? `strlen(string) + 1` could overflow a `uint32_t`, in which case I would expect the function to return `false`
- Computing `strlen(string)` three times :sob:
- The only negative value that `write()` should ever return is `-1`.  I'd usually recommend checking the failure state (negative return value) and checking the length written as two separate expressions -- and casting the returned length written to `size_t` rather than the other way round.

Here's a vague attempt at sanification of this:

``` c++
static bool startprocess_write_cstring_to_fd(int fd, char *string)
{
    size_t t_len = strlen(string) + 1;
    if (len > UINT32_MAX)
        return false;

    if (!startprocess_write_uint32_t_fd(fd, uint32_t(t_len)))
        return false;

    ssize_t t_status = write(fd, string, t_len);

    /* This verifies that t_status is positive, at which point you _know_
     * that it will fit into a size_t losslessly */
    return (t_status >= 0 && size_t(t_status) == t_len);
}
```

peter-b commented on engine/src/exec-strings-chunk.cpp:
```diff
@@ -285,7 +285,7 @@ void MCStringsMarkTextChunkInRange(MCExecContext& ctxt, MCStringRef p_string, MC
                 // If we didn't, and this operation does not force additional delimiters, then include the previous delimiter in the mark.
                 // e.g. mark item 3 of a,b,c -> a,b(,c) so that delete item 3 of a,b,c -> a,b
-                else if (r_start > p_range . offset && !r_add)
+                else if (uindex_t(r_start) > p_range . offset && !r_add)
```
Please add a `r_start >=0` check.

peter-b commented on engine/src/exec-strings-chunk.cpp:
```diff
@@ -355,7 +355,7 @@ void MCStringsMarkTextChunkInRange(MCExecContext& ctxt, MCStringRef p_string, MC
 	            if (r_end >= 0 && (uindex_t) r_end < t_length)
                     r_end++;
-                else if (r_start > p_range . offset && !r_add)
+                else if (uindex_t(r_start) > p_range . offset && !r_add)
```
Please add `r_start >= 0`.

peter-b commented on engine/src/exec-strings-chunk.cpp:
```diff
@@ -415,7 +415,7 @@ void MCStringsMarkTextChunkInRange(MCExecContext& ctxt, MCStringRef p_string, MC
 	            if (r_end >= 0 && (uindex_t) r_end == t_length)
                 {
-                    while (r_start > p_range . offset && MCUnicodeIsWhitespace(MCStringGetCharAtIndex(p_string, r_start - 1)))
+                    while (uindex_t(r_start) > p_range . offset && MCUnicodeIsWhitespace(MCStringGetCharAtIndex(p_string, r_start - 1)))
```
Yep, you guessed it!

peter-b commented:
@montegoulding I think we need to be very careful about adding casts in order to silence signed/unsigned comparisons. I feel that it is often better to leave the warning in as an indication that "something is not quite right here" than to insert a typecast that could hide a real logic error.  For example, in `MCStringsMarkTextChunkInRange()` you cast a signed (`r_start`) to an unsigned -- but without adding a test that it is actually positive.

peter-b commented:
@montegoulding It's interesting to look through your changes and see that there are loads of places where, for example, we store a datum as `uint32_t`, but if it has a value larger than `INT32_MAX` then everything will break.  `MCImageBitmap::width` is a good example.

peter-b commented on revdb/src/postgresql_connection.cpp:
```diff
@@ -451,13 +451,12 @@ void DBConnection_POSTGRESQL::getTables(char *buffer, int *bufsize)
 				unsigned int colsize;
 				char *coldata = newcursor->getFieldDataBinary(1,colsize);
-				colsize = strlen(coldata);
+				colsize = (unsigned int)strlen(coldata);
```
Is there any danger that this could be an unchecked overflow?

montegoulding replied:
Actually I'm not even sure why it's necessary as `getFieldDataBinary` should be setting `colsize`

peter-b commented on lcidlc/src/Parser.cpp:
```diff
@@ -143,6 +143,8 @@ static bool ParserReport(ParserRef self, Position p_where, ParserError p_error,
 			fprintf(stderr, "%s'%s' expected\n", i == 0 ? "" : " or ", p_hints[i]);
 		}
 		break;
+    default:
```
This should be `case kParserErrorNone:`.  You want to get a build warning (or better, a build error) when you add a new type of parser error but don't teach `ParserReport()` to handle it.

peter-b commented:
@montegoulding It's unclear why we need to pass a `-no_compact_unwind` flag to the linker for OS X.  That's the kind of thing that really should be explained in a commit message or even in an inline comment in the .gypi file, please.

peter-b commented on libexternal/src/osxsupport.cpp:
```diff
@@ -39,7 +39,7 @@ char *string_to_utf8(const char *p_string)
 	{
 		unsigned int t_string_length;
-		t_string_length = strlen(p_string);
+		t_string_length = (unsigned int)strlen(p_string);
```
This is an unchecked overflow on 64-bit systems.  It would be better to refactor with `size_t t_string_length;`

peter-b replied:
Actually you may need to explicitly check and fail if `strlen(p_string) > UINDEX_MAX`, I think.

montegoulding replied:
Hmm... what are your thoughts on this which will truncate the string to the maximum size we can fit into our unfortunately guessed buffer size:

```
size_t t_string_length;
t_string_length = strlen(p_string);
if (t_string_length > SSIZE_MAX)
            t_string_length = SSIZE_MAX;
```

livecodefraser commented:
I'm not sure I like fixing the warning by entirely disabling the generation of compact unwind information (it feels like using a sledgehammer to drive in a pin!). Not that I have any better ideas, though...

runrevmark commented:
How about _not_ disabling compact unwind, and just leave the linker to generate them... After all it is not an error if the linker generates a warning, it is information about the final code it managed to generate - i.e. a build should not fail if a compact unwind warning is issued from the linker.

montegoulding commented:
I'll reenable the compact unwind warning if you like. I asked on slack about this and as told the unwind info is only needed for exception handling which we don't do. I had the impression it was just build noise but if it's not then it's not.

Commit:
Revert "Silence compact unwind warning"

This reverts commit 7df64a37f61759497a6298658b93f5defc7956e4.

Commit:
Improvements based on comments

Commit:
Fix BIO_read and BIO_write API use

Commit:
Turn on -Werror=missing-field-initializers

Commit:
Turn on -Werror=missing-declarations

Commit:
Turn on -Werror=unused-variable

Commit:
Fix conversion warnings

Commit:
Only #pragma optimise for __VISUALC__ as only compiler providing support

Commit:
Fix some conversion warnings

Commit:
Resolve some conversion warnings in lc-compile

Commit:
Silence warning on comparison between Chunk_term and anonymous

Commit:
Merge commit 'b05310e88a4741f984997a2f3905da4c32ad3fae' into macwarnings

Commit:
Remove unused variables in mac 32 bit build

Commit:
Fix warnings on use of int for buffer size

Commit:
Use delete[] for pointers allocated with new[]

Commit:
Cast to lower output as char

Commit:
Fix some type conversion warnings

Commit:
Fix implicit conversions of nil

Commit:
Fix implicit type conversion

Commit:
Implicit conversion of nil

Commit:
Fix redefinition of S_ISLNK

montegoulding commented:
It is relatively amazing that there's no merge conflicts on this yet

montegoulding commented:
Now that this has merge conflicts I'll close and cherrypick what I can into separate PRs so it's easier to review
